cluster/etcd_service_discovery.go

```diff
@@ -454,12 +466,14 @@ func (sd *etcdServiceDiscovery) revoke() error {
 
 func (sd *etcdServiceDiscovery) addServer(sv *Server) {
 	if _, loaded := sd.serverMapByID.LoadOrStore(sv.ID, sv); !loaded {
-		mapSvByType, ok := sd.serverMapByType.Load(sv.Type)
+		sd.mapByTypeLock.Lock()
```

Comment:
Defer?

Reply:
Same as above, using defer would keep the lock longer than needed

Reply:
Same as my reply in the first comment about defers.